src/lib/state.ts

```diff
@@ -0,0 +1,277 @@
+type StateAttachedContract<A> = {
+  key: string;
+  stateType: AsFieldElements<A>;
+  instance: SmartContract;
+  class: typeof SmartContract;
+};
+
+type InternalStateType<A> = State<A> & { _contract?: StateAttachedContract<A> };
+
+function createState<T>(): InternalStateType<T> {
+  return {
+    _contract: undefined as StateAttachedContract<T> | undefined,
+
+    set(state: T) {
+      if (this._contract === undefined)
+        throw Error(
+          'set can only be called when the State is assigned to a SmartContract @state.'
+        );
+      let layout = getLayoutPosition(this._contract);
+      let stateAsFields = this._contract.stateType.toFields(state);
+      let party = this._contract.instance.self;
+      stateAsFields.forEach((x, i) => {
+        Party.setValue(party.body.update.appState[layout.offset + i], x);
+      });
+    },
+
+    assertEquals(state: T) {
+      if (this._contract === undefined)
+        throw Error(
+          'assertEquals can only be called when the State is assigned to a SmartContract @state.'
+        );
+      let layout = getLayoutPosition(this._contract);
+      let stateAsFields = this._contract.stateType.toFields(state);
+      let party = this._contract.instance.self;
+
+      stateAsFields.forEach((x, i) => {
+        party.body.preconditions.account.state[layout.offset + i].isSome =
+          Bool(true);
+        party.body.preconditions.account.state[layout.offset + i].value = x;
+      });
+    },
+
+    get() {
+      if (this._contract === undefined)
+        throw Error(
+          'get can only be called when the State is assigned to a SmartContract @state.'
+        );
+      let layout = getLayoutPosition(this._contract);
+      let address: PublicKey = this._contract.instance.address;
+      let stateAsFields: Field[];
+      let inProver = GlobalContext.inProver();
+      if (!GlobalContext.inCompile()) {
+        let account: Account;
+        try {
+          account = Mina.getAccount(address);
+        } catch (err) {
+          // TODO: there should also be a reasonable error here
+          if (inProver) {
+            throw err;
+          }
+          throw Error(
+            `${this._contract.key}.get() failed, because the zkapp account was not found in the cache. ` +
+              `Try calling \`await fetchAccount(zkappAddress)\` first.`
+          );
+        }
+        if (account.zkapp === undefined) {
+          // if the account is not a zkapp account, let the default state be all zeroes
+          stateAsFields = Array(layout.length).fill(Field.zero);
+        } else {
+          stateAsFields = [];
+          for (let i = 0; i < layout.length; ++i) {
+            stateAsFields.push(account.zkapp.appState[layout.offset + i]);
+          }
+        }
+        // in prover, create a new witness with the state values
+        // outside, just return the state values
+        stateAsFields = inProver
```

Comment:
Wouldn’t this make the circuit have different constraints if you are inProver vs outside? The witness statement makes variables, right? But the other side of the ternary doesn’t make any variables.

Reply:
yes exactly. there are three cases which are all handled here:
* in compile, make a variable without a value
* in the prover, get the value and make a variable with the value inside
* outside any circuits, get the value and just return it